### ISSUE TITLE
Model sizes

### ISSUE THREAD

**Issue by joshwa71** [NM512/dreamerv3-torch#2]
Hey, thanks so much for the repo, I was looking for a streamlined pytorch implementation.

I'm looking to increase the model size and I'm wondering which config attributes correspond to which attributes from the Dreamer paper. Most are intuitive but some aren't so much.

Any help appreciated.

**Comment by NM512** [comment]
Modified some attributes in config file to make them easy to understand.
Parameters related to model size and its corresponding name in config are listed below.

**Parameter name in the paper (name in config)**
- GRU recurrent units (dyn_hidden and dyn_deter both)
- CNN multiplier (cnn_depth)
- Dense hidden units (units)
- MLP layers (reward_layers, discount_layers, value_layers and actor_layers)

According to the paper, you can switch model size by just changing these values on all environments expect Minecraft and DMC Proprio environments which require network structure changes.
Default parameters are set as model size S mentioned at app. B in the paper.

**Comment by NM512** [comment]
Related parameters are as follows.

**Name in the paper (name in config)**
- GRU recurrent units (dyn_hidden and dyn_deter both)
- CNN multiplier (cnn_depth)
- Dense hidden units (units)
- MLP layers (reward_layers, discount_layers, value_layers and actor_layers)

Model size is only dependent to these values.
But be careful of what Minecraft and DMC Proprio environments require changes of network structure.

**Comment by joshwa71** [comment]
Am I right in thinking the original implementation uses an mlp for the encoder/decoder and has associated params in the config. I can't see any corresponding params for this implementation, is this dense hidden units?

**Comment by NM512** [comment]
Yes, author's implementation includes MLP for the encoder/decoder.
However inputs given to MLP is restricted at [this line](https://github.com/danijar/dreamerv3/blob/dab3648a28709689a93b799e18edcf4c6d460479/dreamerv3/nets.py#L202) by regular expression matching.
It looks mlp for the encoder/decoder is used only for Minecraft and DMC Proprio environments.
Please check [this line](https://github.com/danijar/dreamerv3/blob/dab3648a28709689a93b799e18edcf4c6d460479/dreamerv3/configs.yaml#L175) in config.

This is why I informed changes of network structure are necessary for testing on these environments. And there is corresponding topic in todo list at readme.

As for your training issue, It's really appreciated if you share details since this implementation is tested on only vision DMC suite so far.

**Comment by NM512** [comment]
Yes, author's implementation includes MLP for the encoder/decoder.
However inputs given to MLP is restricted at [this line](https://github.com/danijar/dreamerv3/blob/dab3648a28709689a93b799e18edcf4c6d460479/dreamerv3/nets.py#L202) by regular expression matching.
It looks MLP for the encoder/decoder is used only for Minecraft and DMC Proprio environments.
Please check [this line](https://github.com/danijar/dreamerv3/blob/dab3648a28709689a93b799e18edcf4c6d460479/dreamerv3/configs.yaml#L175) in config.

This is why I informed changes of network structure are necessary for testing on these environments. And there is corresponding topic in todo list at readme.

As for your training issue, It's really appreciated if you share details since this implementation is tested on only vision DMC suite so far.

**Comment by NM512** [comment]
Yes, author's implementation includes MLP for the encoder/decoder.
However inputs given to MLP is selected at [this line](https://github.com/danijar/dreamerv3/blob/dab3648a28709689a93b799e18edcf4c6d460479/dreamerv3/nets.py#L202) by regular expression matching.
It looks MLP for the encoder/decoder is used only for Minecraft and DMC Proprio environments.
Please check [this line](https://github.com/danijar/dreamerv3/blob/dab3648a28709689a93b799e18edcf4c6d460479/dreamerv3/configs.yaml#L175) in config.

This is why I informed changes of network structure are necessary for testing on these environments. And there is corresponding topic in todo list at readme.

As for your training issue, It's really appreciated if you share details since this implementation is tested only on vision DMC suite so far.

**Comment by joshwa71** [comment]
Thanks for the help and of course:

I'm attempting to train the model on a safety gym environment. So far all I have changed is the image keys of the model (and grad heads). None of the models losses appear to be decreasing significantly, especially the image loss. 

![trainopenl](https://user-images.githubusercontent.com/97907284/226148165-53883127-dd64-45be-83b0-caa6d031a138.jpg)
![trainreturn](https://user-images.githubusercontent.com/97907284/226148166-fd8d98e6-db83-4ccb-9ff6-6d7e203e3c07.jpg)

**Comment by NM512** [comment]
Thank you for sharing your state!

It seems at least image reconstruction is processed successfully because images of top row and middle row look similar.
Please check around wrapper.py for confirming interface like observation/action/reward between the algorithm and your environment are as you intended.

It is useful to run your environment with author's implementation and compare each statistics with mine in tensorboard log.

**Comment by NM512** [comment]
Thank you for sharing your state!

It seems at least image reconstruction is processed successfully because images of top row and middle row look similar.
Please check around wrapper.py for confirming interface like observation/action/reward between the algorithm and your environment are as you intended.

It is useful to run your environment with author's implementation and compare each log with mine in tensorboard.

**Comment by NM512** [comment]
Thank you for sharing your state!

It seems at least image reconstruction is processed successfully because images of top row and middle row look similar.
Please check around wrapper.py for confirming interface like observation/action/reward between the algorithm and your environment are as you intended.

It would be very helpful if I could run the environment with the author's implementation and compare each log with mine in tensorboard.

**Comment by NM512** [comment]
Thank you for sharing your state!

It seems at least image reconstruction is processed successfully because images of top row and middle row look similar.
Please check around wrapper.py for confirming interface like observation/action/reward between the algorithm and your environment are as you intended.

It would be very helpful if you could run the environment with the author's implementation and compare each log with mine in tensorboard.

**Comment by NM512** [comment]
Thank you for sharing your state!
It seems at least image reconstruction is processed successfully because images of top row(ground truth) and middle row(predicted) look similar.

Have you editted wrapper.py for making observation/action/reward  as you intended?

**Comment by joshwa71** [comment]
Yes I have made some minor edits to wrapper.py. I'm using a safety gym env and so have included the cost in the episode logging so I can track the cost as the algorithm learns (though I haven't added cost to the reward function or model itself). After testing the original repo on the same env I got great results, with learning starting almost immediately.

The safety gym env is wrapped in a PixelObservationWrapper which returns a dictionary with key "image" after each step.

If the image reconstruction is working as intended, it must be some issue with learning in imagination, no?

**Comment by NM512** [comment]
hmmm...
There might be some issue with learning in imagination or some component like reward head.
I can take a look if you share tensorboard logs of author's repository and this repository.
It is also helpful if you can share the code.

**Comment by NM512** [comment]
hmmm...
There might be some issue with learning in imagination or some component like reward head.
I can take a look if you share tensorboard logs of author's repository and this repository.
It is also helpful if you can share the code.

Could you also confirm you are using latest version of this repository?

**Comment by joshwa71** [comment]
I have created a repo of the modified code so you can see my changes:

**Comment by joshwa71** [comment]
I have created a repo of the modified code so that you can see my changes:
https://github.com/joshwa71/SafetyAwayModelBasedRL

Also, here are some of the relevant training metrics from training on the original env:


![metrics4](https://user-images.githubusercontent.com/97907284/226174345-567ee89d-0d5b-4097-b3fb-44160aa28c0e.jpg)
![metrics3](https://user-images.githubusercontent.com/97907284/226174346-5b78d3b1-f501-48d5-8ff0-2497f8e20fcd.jpg)
![metrics2](https://user-images.githubusercontent.com/97907284/226174349-3f64fdc7-ffa0-4c8f-b84d-d75a2ed846c5.jpg)
![Metrics1](https://user-images.githubusercontent.com/97907284/226174350-9552ce44-de65-48c9-9f00-bc4c0b472a22.jpg)

**Comment by NM512** [comment]
Thanks for sharing the code.
I noticed you are using older code which includes some bugs.
After updating it looks learning.

![train_return](https://user-images.githubusercontent.com/70328564/226180088-c1f9c211-0392-41a3-ad17-711ae0027aaa.png)

This is GIF image at step 15000
![eval](https://user-images.githubusercontent.com/70328564/226180139-dfa169bd-b35d-4b6e-b025-2ae017ddbb81.gif)

I uploaded used code and tensorboard log here.
https://github.com/NM512/SafetyAwayModelBasedRL

Enjoy Safety Gym!! :D

**Comment by joshwa71** [comment]
This repo works perfectly, thank you!

May I ask what changes you made, as I spent a little while trying to figure it out.

Thanks again.

**Comment by NM512** [comment]
I'm grad to hear that.
Following 2 factors had strongest effect against performance when I tested on Vision DMC suite.
1. How to initialize network weights
2. How to calculate loss of two hot distribution
Please take a look code before and after of [this commit](https://github.com/NM512/dreamerv3-torch/commit/627344439447ec2394ecafce0ba85a1798edb47b#diff-2171db0cf84575462627421658645f2d76c71fb0e13dceb297c41e443bfd8af7L688) if you are interested in.

**Comment by NM512** [comment]
I'm grad to hear that.
I made many lines of modification.
And following 2 factors had strongest effect against performance when I tested on Vision DMC suite.
1. How to initialize network weights
2. How to calculate loss of two hot distribution

Please take a look the code before and after of [this commit](https://github.com/NM512/dreamerv3-torch/commit/627344439447ec2394ecafce0ba85a1798edb47b#diff-2171db0cf84575462627421658645f2d76c71fb0e13dceb297c41e443bfd8af7L688) if you are interested in.

**Comment by NM512** [comment]
I'm grad to hear that. I made many lines of modification.
In that changes, following 2 factors had strongest effect against performance when I tested on Vision DMC suite.
1. How to initialize network weights
2. How to calculate loss of two hot distribution

Please take a look code before and after of [this commit](https://github.com/NM512/dreamerv3-torch/commit/627344439447ec2394ecafce0ba85a1798edb47b#diff-2171db0cf84575462627421658645f2d76c71fb0e13dceb297c41e443bfd8af7L688) if you are interested in.

**Comment by stablegradients** [comment]
Could you provide the yml files with the exact values for the xs, s, m, L, and xl models. This would be super helpful just to make sure we have not made a mistake in the runs.
